README.md
citecorp
=========



[![cran checks](https://cranchecks.info/badges/worst/citecorp)](https://cranchecks.info/pkgs/citecorp)
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Build Status](https://travis-ci.com/ropenscilabs/citecorp.svg?branch=master)](https://travis-ci.com/ropenscilabs/citecorp)
[![rstudio mirror downloads](https://cranlogs.r-pkg.org/badges/citecorp)](https://github.com/metacran/cranlogs.app)
[![cran version](https://www.r-pkg.org/badges/version/citecorp)](https://cran.r-project.org/package=citecorp)

Client for the Open Citations Corpus http://opencitations.net/ (OCC)

OCC created their own identifiers called Open Citation Identifiers (oci), e.g., 

```
020010009033611182421271436182433010601-02001030701361924302723102137251614233701000005090307
```

You are probably not going to be using oci identifiers, but rather DOIs and/or PMIDs
and/or PMCIDs. See `?oc_lookup` for methods for cross-walking among identifier types.

If you'd like to use the OpenCitations Sparql endpoint yourself you can find that
at http://opencitations.net/sparql


## Install

CRAN version


```r
install.packages("citecorp")
```

Development version


```r
remotes::install_github("ropenscilabs/citecorp")
```


```r
library("citecorp")
```

## Methods for converting IDs


```r
oc_doi2ids("10.1097/igc.0000000000000609")
#>                            doi                           paper      pmcid
#> 1 10.1097/igc.0000000000000609 https://w3id.org/oc/corpus/br/1 PMC4679344
#>       pmid
#> 1 26645990
oc_pmid2ids("26645990")
#>                            doi                           paper      pmcid
#> 1 10.1097/igc.0000000000000609 https://w3id.org/oc/corpus/br/1 PMC4679344
#>       pmid
#> 1 26645990
oc_pmcid2ids("PMC4679344")
#>                            doi                           paper      pmcid
#> 1 10.1097/igc.0000000000000609 https://w3id.org/oc/corpus/br/1 PMC4679344
#>       pmid
#> 1 26645990
```

You can pass in more than one identifer to each of the above functions:


```r
oc_doi2ids(oc_dois[1:6])
#>                                  doi                                 paper
#> 1               10.1128/jvi.00758-10 https://w3id.org/oc/corpus/br/5357460
#> 2 10.1111/j.2042-3306.1989.tb02167.x  https://w3id.org/oc/corpus/br/589891
#> 3       10.1097/rli.0b013e31821eea45 https://w3id.org/oc/corpus/br/3931705
#> 4           10.1177/0148607114529597 https://w3id.org/oc/corpus/br/5016780
#> 5            10.1111/1567-1364.12217 https://w3id.org/oc/corpus/br/3819297
#> 6      10.1016/s0168-9525(99)01798-9 https://w3id.org/oc/corpus/br/4606537
#>        pmcid     pmid
#> 1 PMC2953162 20702630
#> 2       <NA>  2670542
#> 3       <NA> 21577119
#> 4       <NA> 24711119
#> 5       <NA> 25263709
#> 6       <NA> 10461200
```

## COCI methods

OpenCitations Index of Crossref open DOI-to-DOI references

If you don't load `tibble` you get normal data.frame's


```r
library(tibble)
doi1 <- "10.1108/jd-12-2013-0166"
# references
oc_coci_refs(doi1)
#> # A tibble: 37 x 7
#>    cited         timespan citing    journal_sc creation oci            author_sc
#>  * <chr>         <chr>    <chr>     <chr>      <chr>    <chr>          <chr>    
#>  1 10.1001/jama… P9Y2M5D  10.1108/… no         2015-03… 0200101000836… no       
#>  2 10.1002/asi.… P41Y8M   10.1108/… no         2015-03… 0200101000836… no       
#>  3 10.1002/(sic… P25Y6M   10.1108/… no         2015-03… 0200101000836… no       
#>  4 10.1007/bf02… P17Y2M   10.1108/… no         2015-03… 0200101000836… no       
#>  5 10.1007/s105… P2Y2M3D  10.1108/… no         2015-03… 0200101000836… no       
#>  6 10.1007/s111… P5Y8M27D 10.1108/… no         2015-03… 0200101000836… no       
#>  7 10.1016/j.we… P2Y3M    10.1108/… no         2015-03… 0200101000836… no       
#>  8 10.1016/j.we… P1Y10M   10.1108/… no         2015-03… 0200101000836… no       
#>  9 10.1023/a:10… P12Y     10.1108/… no         2015-03… 0200101000836… no       
#> 10 10.1038/3507… P13Y10M  10.1108/… no         2015-03… 0200101000836… no       
#> # … with 27 more rows
# citations
oc_coci_cites(doi1)
#> # A tibble: 20 x 7
#>    cited     timespan  citing    journal_sc creation oci               author_sc
#>  * <chr>     <chr>     <chr>     <chr>      <chr>    <chr>             <chr>    
#>  1 10.1108/… P3Y       10.1145/… no         2018     0200101040536030… no       
#>  2 10.1108/… P2Y5M     10.1057/… no         2017-08  0200100050736280… no       
#>  3 10.1108/… P4Y1M1D   10.3233/… no         2019-04… 0200302030336132… no       
#>  4 10.1108/… P4Y5M10D  10.3233/… no         2019-08… 0200302030336132… no       
#>  5 10.1108/… P1Y0M14D  10.3233/… no         2016-03… 0200302030336283… no       
#>  6 10.1108/… P3Y10M12D 10.3233/… no         2019-01… 0200302030336283… no       
#>  7 10.1108/… P3Y6M     10.1142/… no         2018-09  0200101040236280… no       
#>  8 10.1108/… P2Y11M20D 10.7554/… no         2018-03… 0200705050436142… no       
#>  9 10.1108/… P0Y       10.3346/… no         2015     0200303040636192… no       
#> 10 10.1108/… P3Y       10.1007/… no         2018     0200100000736090… no       
#> 11 10.1108/… P2Y3M4D   10.1093/… no         2017-06… 0200100090336111… no       
#> 12 10.1108/… P2Y       10.1007/… no         2017     0200100000736090… no       
#> 13 10.1108/… P3Y11M    10.1186/… no         2019-02  0200101080636280… no       
#> 14 10.1108/… P3Y       10.1007/… no         2018     0200100000736090… no       
#> 15 10.1108/… P3Y       10.1007/… no         2018     0200100000736090… no       
#> 16 10.1108/… P0Y       10.1007/… no         2015     0200100000736090… no       
#> 17 10.1108/… P2Y       10.1007/… no         2017     0200100000736090… no       
#> 18 10.1108/… P2Y9M12D  10.1007/… no         2017-12… 0200100000736090… no       
#> 19 10.1108/… P1Y11M5D  10.1101/… no         2017-02… 0200101000136010… no       
#> 20 10.1108/… P1Y4M1D   10.1177/… no         2016-07… 0200101070736000… no
# metadata
oc_coci_meta(doi1)
#> # A tibble: 1 x 13
#>   issue oa_link citation_count author source_title citation source_id volume
#> * <chr> <chr>   <chr>          <chr>  <chr>        <chr>    <chr>     <chr> 
#> 1 2     http:/… 20             Peron… Journal Of … 10.1145… issn:002… 71    
#> # … with 5 more variables: year <chr>, page <chr>, doi <chr>, reference <chr>,
#> #   title <chr>
```


## Meta

* Please [report any issues or bugs](https://github.com/ropenscilabs/citecorp/issues)
* License: MIT
* Get citation information for `citecorp` in R doing `citation(package = 'citecorp')`
* Please note that this project is released with a [Contributor Code of Conduct][coc].
By participating in this project you agree to abide by its terms.

[sparqldsl]: https://github.com/ropenscilabs/sparqldsl
[coc]: https://github.com/ropenscilabs/citecorp/blob/master/CODE_OF_CONDUCT.md


[![rofooter](https://ropensci.org/public_images/github_footer.png)](https://ropensci.org)


NEWS.md
citecorp 0.3.0
==============

### NEW FEATURES

* `oc_coci_refs()`, `oc_coci_cites()`, and `oc_coci_citation()` now support more than one DOI or COCI identifier passed in. `oc_coci_meta()` already supported more than one identifier; now all four `oc_coci*` methods support more than one identifier. The caveat is that the API route behindn `oc_coci_meta()` supports >1 identifier in a single http request - whereas the API routes behind the other functions only support one identifier per request, so they make an http request for each identifier passed in (#6)

### MINOR IMPROVEMENTS

* added tests for Cloudflare http status codes (#7)


citecorp 0.2.2
==============

### BUG FIXES

* better check if http requests will work before running examples that require http requests (#5)


citecorp 0.2.0
==============

### NEW FEATURES

* the functions `oc_doi2ids`, `oc_pmid2ids`, and `oc_pmcid2ids` (see `?oc_lookup`) now all accept more than 1 identifier. the underlying SPARQL queries were modified to make this possible. (#1)
* three new package datasets (oc_dois, oc_pmids, oc_pmcids) with vectors of identifiers of each of those types for testing and examples

### BUG FIXES

* fixed an error in parsing results in `oc_lookup` functions - this fix was later changed as part of another fix, but still appreciated (#2) (#3) thanks @Selbosh
* fix to parsing in `oc_lookup` functions: only retrieve data.frame columns if they exist - prevents error on retrieving a column that doesn't exist (#4) thanks @Selbosh


citecorp 0.1.0
==============

### NEW FEATURES

* Released to CRAN


tests/test-all.R
library("testthat")
test_check("citecorp")


tests/testthat/test-oc_coci.R
doi1 <- "10.1108/jd-12-2013-0166"
doi2 <- "10.6084/m9.figshare.3443876"
doi3 <- "10.1371/journal.pgen.1005937"
oci1 <-
  "02001010806360107050663080702026306630509-0200101080636102704000806"
oci2 <-
  "0200101000836191363010263020001036300010606-020010003083604090301050910"

context("oc_coci_refs")
test_that("oc_coci_refs works", {
  vcr::use_cassette("oc_coci_refs", {
    x <- oc_coci_refs(doi1)
  })

  expect_is(x, "data.frame")
  expect_is(x, "tbl_df")
  expect_is(x, "tbl")
  expect_is(x$cited, "character")
  expect_is(x$oci, "character")
  expect_is(x$citing, "character")

  # all DOIs
  expect_match(x$cited, "10\\.")
  expect_match(x$citing, "10\\.")
  
  # all citing equal DOI passed in
  expect_equal(length(unique(x$citing)), 1)
  # all cited are other DOIs
  expect_false(any(grepl(doi1, x$cited)))
})
test_that("oc_coci_refs - many dois", {
  vcr::use_cassette("oc_coci_refs_many_dois", {
    x <- oc_coci_refs(c(doi1, doi3))
  })
  expect_equal(sort(unique(x$citing)), sort(c(doi1, doi3)))
})
test_that("oc_coci_refs fails well", {
  expect_error(oc_coci_refs(), "argument \"doi\" is missing")
  expect_error(oc_coci_refs(5), "doi must be of class")
})



context("oc_coci_cites")
test_that("oc_coci_cites works", {
  vcr::use_cassette("oc_coci_cites", {
    x <- oc_coci_cites(doi1)
  })

  expect_is(x, "data.frame")
  expect_is(x, "tbl_df")
  expect_is(x, "tbl")
  expect_is(x$cited, "character")
  expect_is(x$oci, "character")
  
  # all DOIs
  expect_match(x$cited, "10\\.")
  expect_match(x$citing, "10\\.")
  
  # all cited are other DOIs
  expect_equal(length(unique(x$cited)), 1)
  # all citing equal DOI passed in
  expect_false(any(grepl(doi1, x$citing)))
})
test_that("oc_coci_cites - many dois", {
  vcr::use_cassette("oc_coci_cites_many_dois", {
    x <- oc_coci_cites(c(doi1, doi3))
  })
  expect_equal(sort(unique(x$cited)), sort(c(doi1, doi3)))
})
test_that("oc_coci_cites fails well", {
  expect_error(oc_coci_cites(), "argument \"doi\" is missing")
  expect_error(oc_coci_cites(5), "doi must be of class")
})



context("oc_coci_meta")
test_that("oc_coci_meta works", {
  vcr::use_cassette("oc_coci_meta", {
    x <- oc_coci_meta(doi1)
  })

  expect_is(x, "data.frame")
  expect_is(x, "tbl_df")
  expect_is(x, "tbl")
  expect_is(x$citation, "character")
  expect_is(x$author, "character")
  expect_is(x$reference, "character")
  
  # all DOIs
  expect_match(x$citation, "10\\.")
  expect_match(x$reference, "10\\.")

  # matches passed in DOI
  expect_match(x$doi, doi1)
})
test_that("oc_coci_meta - many dois", {
  vcr::use_cassette("oc_coci_meta_many_dois", {
    x <- oc_coci_meta(c(doi1, doi3))
  })
  expect_equal(NROW(x), 2)
})
test_that("oc_coci_meta fails well", {
  expect_error(oc_coci_meta(), "argument \"doi\" is missing")
  expect_error(oc_coci_meta(5), "doi must be of class")
})



context("oc_coci_citation")
test_that("oc_coci_citation works", {
  vcr::use_cassette("oc_coci_citation", {
    x <- oc_coci_citation(oci1)
  })

  expect_is(x, "data.frame")
  expect_is(x, "tbl_df")
  expect_is(x, "tbl")
  expect_is(x$cited, "character")
  expect_is(x$oci, "character")
  expect_is(x$citing, "character")

  expect_equal(NROW(x), 1)
  
  # all DOIs
  expect_match(x$cited, "10\\.")
  expect_match(x$citing, "10\\.")
  
  # all cited are other DOIs
  expect_equal(length(unique(x$cited)), 1)
  
  # citing equals OCI passed in
  expect_match(oci1, x$oci)
})
test_that("oc_coci_citation - many dois", {
  vcr::use_cassette("oc_coci_citation_many_dois", {
    x <- oc_coci_citation(c(oci1, oci2))
  })
  expect_equal(NROW(x), 2)
})
test_that("oc_coci_citation fails well", {
  expect_error(oc_coci_citation(), "argument \"oci\" is missing")
  expect_error(oc_coci_citation(5), "oci must be of class")
})


context("cloudflare errors")
test_that("citecorp catches cloudflare errors", {
  webmockr::enable()
  
  stub=webmockr::stub_request("get",
    "http://opencitations.net/index/coci/api/v1/citations/10.1108/jd-12-2013-0166")
  webmockr::to_return(stub, status = 520)
  expect_error(oc_coci_cites(doi1), class = "http_520")

  stub2=webmockr::stub_request("get",
    "http://opencitations.net/index/coci/api/v1/citations/10.1371/journal.pgen.1005937")
  webmockr::to_return(stub2, status = 524)
  expect_error(oc_coci_cites(doi3), class = "http_524")

  webmockr::disable()
})


tests/testthat/test-oc_lookup.R
cols <- sort(c("doi", "pmid", "pmcid", "paper"))

context("oc_doi2ids")
test_that("oc_doi2ids works", {
  vcr::use_cassette("oc_doi2ids", {
    x <- oc_doi2ids("10.1097/igc.0000000000000609")
  })

  expect_is(x, "data.frame")
  expect_equal(sort(names(x)), cols)
  for (i in seq_len(NCOL(x))) expect_is(x[,i], "character")
})

test_that("oc_doi2ids works", {
  vcr::use_cassette("oc_doi2ids_fewer_columns", {
    x <- oc_doi2ids('10.1093/biomet/80.3.527')
    y <- oc_doi2ids('10.1093/biomet/79.3.531')
  })

  expect_is(x, "data.frame")
  expect_equal(sort(names(x)), c("doi", "paper"))
  for (i in seq_len(NCOL(x))) expect_is(x[,i], "character")

  expect_is(y, "data.frame")
  expect_equal(sort(names(y)), c("doi", "paper"))
  for (i in seq_len(NCOL(y))) expect_is(y[,i], "character")
})

test_that("oc_doi2ids fails well", {
  expect_error(oc_doi2ids(), "argument \"id\" is missing")
  expect_error(oc_doi2ids(matrix()), "id must be of class")
})



context("oc_pmid2ids")
test_that("oc_pmid2ids works", {
  vcr::use_cassette("oc_pmid2ids", {
    x <- oc_pmid2ids("26645990")
  })

  expect_is(x, "data.frame")
  expect_equal(sort(names(x)), cols)
  for (i in seq_len(NCOL(x))) expect_is(x[,i], "character")
})

test_that("oc_pmid2ids fails well", {
  expect_error(oc_pmid2ids(), "argument \"id\" is missing")
  expect_error(oc_pmid2ids(matrix()), "id must be of class")
})



context("oc_pmcid2ids")
test_that("oc_pmcid2ids works", {
  vcr::use_cassette("oc_pmcid2ids", {
    x <- oc_pmcid2ids("PMC4679344")
  })

  expect_is(x, "data.frame")
  expect_equal(sort(names(x)), cols)
  for (i in seq_len(NCOL(x))) expect_is(x[,i], "character")
})

test_that("oc_pmcid2ids fails well", {
  expect_error(oc_pmcid2ids(), "argument \"id\" is missing")
  expect_error(oc_pmcid2ids(matrix()), "id must be of class")
})


R/http.R
coci_base <- "http://opencitations.net"

errs <- function(x) {
  if (x$status_code > 201) {
    fun <- fauxpas::find_error_class(x$status_code)$new()
    fun$do_verbose(x)
  }
}

oc_GET <- function(base, path = NULL, args = list(), ...) {
  cli <- crul::HttpClient$new(
    base,
    headers = list("User-Agent" = oc_ua(), "X-USER-AGENT" = oc_ua()),
    opts = list(...)
  )
  res <- cli$get(path, query = args)
  errs(res)
  res$parse("UTF-8")
}

oc_ua <- function() {
  versions <- c(
    paste0("r-curl/", utils::packageVersion("curl")),
    paste0("crul/", utils::packageVersion("crul")),
    sprintf("rOpenSci(citecorp/%s)",
            utils::packageVersion("citecorp"))
  )
  paste0(versions, collapse = " ")
}

oc_sparql_GET <- function(url, path, query, flatten = FALSE, ...) {
  cli <- crul::HttpClient$new(
    url = url, opts = list(...),
    headers = list(
      Accept = "application/sparql-results+json",
      "User-Agent" = oc_ua(),
      "X-USER-AGENT" = oc_ua()
    )
  )
  qry <- list(query = query, format = "json")
  res <- cli$get(path, query = qry)
  res$raise_for_status()
  jsonlite::fromJSON(res$parse("UTF-8"), flatten = flatten)
}

oc_sparql_OK <- function(query) {
  cli <- crul::HttpClient$new(
    url = coci_base, opts = list(timeout_ms = 1000L),
    headers = list(
      Accept = "application/sparql-results+json",
      "User-Agent" = oc_ua(),
      "X-USER-AGENT" = oc_ua()
    )
  )
  url <- cli$url_fetch("sparql", query = list(query = query, format = "json"))
  crul::ok(url)
}

cp_query <- function(query, flatten = TRUE, ...) {
  oc_sparql_GET(coci_base, "sparql", query, flatten = flatten, ...)
}


R/oc_lookup.R
oc_2ids_template <- function(template_string, id_name) {
  function(id, ...) {
    assert(id, c("character", "integer", "numeric"))
    qry <- sprintf(template_string, paste0(sprintf("\"%s\"", id), collapse = " "))
    tmp <- cp_query(qry, ...)$results$bindings
    if (length(tmp) == 0) return(data.frame(NULL))
    df <- tmp[-grep('\\.type', names(tmp))]
    names(df) <- gsub("\\.value", "", names(df))
    names(df)[1] <- id_name
    df <- df[, names(df) %in% c("doi", "pmid", "pmcid", "paper")]
    df <- df[, sort(names(df))]
    return(df)
  }
}

#' Methods for getting IDs from other IDs
#'
#' @export
#' @name oc_lookup
#' @param id One or more digital object identifiers (DOI),
#' PMID, or PMCID, depending on the function
#' @return data.frame, with four columns:
#' 
#' - doi: digital object identifier
#' - pmid: pubmed identifier
#' - pmcid: pubmed central identifier
#' - paper: open citations corpus url
#' 
#' An empty data.frame (no columns or rows) when no results found
#' 
#' Column order will always be the same; note though that some
#' columns may be missing if, for example, there's no PMID for
#' a DOI search.
#' @param ... curl options passed on to [crul::verb-GET]
#' @examples
#' if (oc_lookup_check()) {
#' try(
#'   oc_doi2ids("10.1097/igc.0000000000000609", timeout_ms=10),
#'   silent = TRUE
#' )
#' }
#' 
#' ### More examples
#' \dontrun{
#' oc_doi2ids('10.1093/biomet/80.3.527')
#' oc_doi2ids('10.1093/biomet/79.3.531')
#' oc_pmid2ids("26645990")
#' oc_pmcid2ids("PMC4679344")
#' 
#' oc_doi2ids(id = oc_dois[1:3])
#' oc_pmid2ids(id = oc_pmids[1:3])
#' oc_pmcid2ids(id = oc_pmcids[1:3])
#' }

#' @export
#' @rdname oc_lookup
oc_doi2ids <- oc_2ids_template(qry_doi2ids, "doi")

#' @export
#' @rdname oc_lookup
oc_pmid2ids <- oc_2ids_template(qry_pmid2ids, "pmid")

#' @export
#' @rdname oc_lookup
oc_pmcid2ids <- oc_2ids_template(qry_pmcid2ids, "pmcid")

#' oc_lookup checker
#' @export
#' @keywords internal
oc_lookup_check <- function() {
  qry <- sprintf(qry_doi2ids, "\"10.1093/biomet/80.3.527\"")
  oc_sparql_OK(qry)
}


R/open_citations.R
qry_doi2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?pmid ?pmcid WHERE {
    ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:doi ;
        literal:hasLiteralValue "%s"
    ] .
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmid ;
        literal:hasLiteralValue ?pmid
      ] .
    }
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmcid ;
        literal:hasLiteralValue ?pmcid
      ] .
    }
  }'

qry_pmid2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?doi ?pmcid WHERE {
    ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmid ;
        literal:hasLiteralValue "%s"
    ] .
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:doi ;
        literal:hasLiteralValue ?doi
      ] .
    }
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmcid ;
        literal:hasLiteralValue ?pmcid
      ] .
    }
  }'

qry_pmcid2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?doi ?pmid WHERE {
    ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmcid ;
        literal:hasLiteralValue "%s"
    ] .
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:doi ;
        literal:hasLiteralValue ?doi
      ] .
    }
    
    OPTIONAL {
      ?paper datacite:hasIdentifier [ 
        datacite:usesIdentifierScheme datacite:pmid ;
        literal:hasLiteralValue ?pmid
      ] .
    }
  }'

oc_2ids_template <- function(template_string, id_name) {
  function(id, ...) {
   qry <- sprintf(template_string, id)
    conn <- sparqldsl::SparqlClient$new(
      url = "opencitations.net", path = "sparql")
    tmp <- conn$query(qry, ...)$results$bindings
    if (length(tmp) == 0) return(data.frame(NULL))
    tmp <- data.frame(
      type = gsub("\\.type", "", names(tmp[, grep("\\.type", names(tmp))])),
      value = unlist(unname(c(tmp[, grep("\\.value", names(tmp))]))),
      stringsAsFactors = FALSE
    )
    rbind(tmp, c(id_name, id))
  }
}

#' OpenCitations methods for getting IDs from other IDs
#' 
#' @export
#' @param id Either a digital object identifier (DOI), 
#' PMID, or PMCID, depending on the function
#' @return data.frame, with two columns, one for identifier type
#' and the other for the identifier value
#' @param ... curl options passed on to [crul::HttpClient]
#' @examples
#' oc_doi2ids("10.1097/igc.0000000000000609")
#' oc_pmid2ids("26645990")
#' oc_pmcid2ids("PMC4679344")

#' @export
#' @rdname open_citations
oc_doi2ids <- oc_2ids_template(qry_doi2ids, "doi")

#' @export
#' @rdname open_citations
oc_pmid2ids <- oc_2ids_template(qry_pmid2ids, "pmid")

#' @export
#' @rdname open_citations
oc_pmcid2ids <- oc_2ids_template(qry_pmcid2ids, "pmcid")


R/sparql_strings.R
qry_doi2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?pmid ?pmcid ?value
  WHERE {
    {
      ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:doi ;
          literal:hasLiteralValue ?value
      ] .
      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmid ;
          literal:hasLiteralValue ?pmid
        ] .
      }
      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmcid ;
          literal:hasLiteralValue ?pmcid
        ] .
      }
    } .
    VALUES ?value { %s }
  }'

qry_pmid2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?doi ?pmcid ?value
  WHERE {
    {
      ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmid ;
          literal:hasLiteralValue ?value
      ] .

      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:doi ;
          literal:hasLiteralValue ?doi
        ] .
      }

      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmcid ;
          literal:hasLiteralValue ?pmcid
        ] .
      }
    } .
    VALUES ?value { %s }
  }'

qry_pmcid2ids <- 'PREFIX datacite: <http://purl.org/spar/datacite/>
  PREFIX literal: <http://www.essepuntato.it/2010/06/literalreification/>
  SELECT ?paper ?doi ?pmid ?value
  WHERE {
    {
      ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmcid ;
          literal:hasLiteralValue ?value
      ] .

      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:doi ;
          literal:hasLiteralValue ?doi
        ] .
      }

      OPTIONAL {
        ?paper datacite:hasIdentifier [
          datacite:usesIdentifierScheme datacite:pmid ;
          literal:hasLiteralValue ?pmid
        ] .
      }
    } .
    VALUES ?value { %s }
  }'


R/citecorp-package.R
#' @title citecorp
#' @description Client for the Open Citations Corpus http://opencitations.net/
#' @importFrom fauxpas http find_error_class
#' @importFrom data.table setDF rbindlist
#' @importFrom jsonlite fromJSON
#' @importFrom crul HttpClient
#' @name citecorp-package
#' @aliases citecorp
#' @docType package
#' @author Scott Chamberlain \email{myrmecocystus@@gmail.com}
#' @keywords package
NULL

#' Vector of 25 DOIs (Digital Object Identifiers)
#'
#' To be used in examples, etc. for [oc_lookup] and [oc_coci]
#' methods
#'
#' @name oc_dois
#' @docType data
#' @keywords data
NULL

#' Vector of 24 PMIDs (PubMed Identifiers)
#'
#' To be used in examples, etc. for [oc_lookup] and [oc_coci]
#' methods
#'
#' @name oc_pmids
#' @docType data
#' @keywords data
NULL

#' Vector of 8 PMCIDs (PubMed Central Identifiers)
#'
#' To be used in examples, etc. for [oc_lookup] and [oc_coci]
#' methods
#'
#' @name oc_pmcids
#' @docType data
#' @keywords data
NULL


R/oc_coci.R
#' COCI: OpenCitations Index of Crossref open
#' DOI-to-DOI references
#'
#' AFAICT this API is a REST wrapper around the SPARQL service
#'
#' @name oc_coci
#' @param doi (character) one or more Digital Object Identifiers (DOIs)
#' @param oci (character) one or more Open Citation Identifiers (OCIs)
#' @param exclude (character) a field_name; all the rows that have an empty
#' value in the field_name specified are removed from the result set
#' @param filter `=<field_name>:<operator><value>:` only the rows compliant
#' with `<value>` are kept in the result set. The parameter `<operation>`
#' is not mandatory. If `<operation>` is not specified, `<value>` is
#' interpreted as a regular expression, otherwise it is compared by means
#' of the specified operation. Possible operators are "=", "<", and ">".
#' For instance, `filter=title:semantics?` returns all the rows that
#' contain the string "semantic" or "semantics" in the field title,
#' while `filter=date:>2016-05` returns all the rows that have a date
#' greater than May 2016.
#' @param sort `=<order>(<field_name>):` sort in ascending (`<order>` set to
#' "asc") or descending (`<order>` set to "desc") order the rows in the
#' result set according to the values in `<field_name>`. For instance,
#' `sort=desc(date)` sorts all the rows according to the value specified
#' in the field date in descending order.
#' @param ... curl options passed on to [crul::verb-GET]
#' @return data.frame, see http://opencitations.net/index/coci/api/v1 for
#' explanation of the resulting columns
#' @references http://opencitations.net/index/coci/api/v1,
#' https://github.com/opencitations/api-coci
#' @examples
#' doi1 <- "10.1108/jd-12-2013-0166"
#' doi2 <- "10.1371/journal.pgen.1005937"
#' oci1 <-
#'  "02001010806360107050663080702026306630509-0200101080636102704000806"
#' oci2 <-
#'  "0200101000836191363010263020001036300010606-020010003083604090301050910"
#' 
#' if (
#' crul::ok(
#' "http://opencitations.net/index/coci/api/v1/references/10.1108/jd-12-2013-0166",
#' timeout_ms = 1000L)
#' ) {
#' try(
#'   oc_coci_cites(doi1),
#'   silent = TRUE
#' )
#' }
#' 
#' ### More examples
#' \dontrun{
#' # references
#' oc_coci_refs(doi1, exclude = "oci")
#' oc_coci_refs(doi1, filter = "date:>2016-05", verbose = TRUE)
#' oc_coci_refs(doi2)
#' oc_coci_refs(c(doi1, doi2))
#'
#' # citations
#' oc_coci_cites(doi1, exclude = "oci")
#' oc_coci_cites(doi2)
#' oc_coci_cites(c(doi1, doi2))
#'
#' # metadata
#' oc_coci_meta(doi2)
#' oc_coci_meta(c(doi1, doi2))
#'
#' # citation - an OCI instead of a DOI
#' oc_coci_citation(oci1)
#' oc_coci_citation(c(oci1, oci2))
#' }

#' @export
#' @rdname oc_coci
oc_coci_refs <- function(doi, exclude=NULL, filter=NULL, sort=NULL, ...) {
  lst2df_tbl(lapply(doi, function(w)
    oc_coci_stub("references", w, exclude, filter, sort, ...)))
}

#' @export
#' @rdname oc_coci
oc_coci_cites <- function(doi, exclude=NULL, filter=NULL, sort=NULL, ...) {
  lst2df_tbl(lapply(doi, function(w)
    oc_coci_stub("citations", w, exclude, filter, sort, ...)))
}

#' @export
#' @rdname oc_coci
oc_coci_meta <- function(doi, exclude=NULL, filter=NULL, sort=NULL, ...) {
  oc_coci_stub("metadata", doi, exclude, filter, sort, ...)
}

#' @export
#' @rdname oc_coci
oc_coci_citation <- function(oci, ...) {
  assert(oci, "character")
  lst2df_tbl(lapply(oci, oc_coci_citation_one, ...))
}

# helpers
oc_coci_citation_one <- function(x, ...) {
  z <- oc_GET(coci_base, file.path(coci_meta, "citation", x), ...)
  oc_coci_parser(z)
}

oc_coci_stub <- function(path, doi, exclude, filter, sort, ...) {
  assert(doi, "character")
  assert(exclude, "character")
  assert(filter, "character")
  assert(sort, "character")
  args <- cp(list(exclude = exclude, filter = filter, sort = sort))
  doi <- paste0(doi, collapse = "__")
  x <- oc_GET(coci_base, file.path(coci_meta, path, doi),
    args = args, ...)
  oc_coci_parser(x)
}

oc_coci_parser <- function(x) {
  x <- jsonlite::fromJSON(x, FALSE)
  structure(lst2df(x), class = c("tbl_df", "tbl", "data.frame"))
}

coci_meta <- "index/coci/api/v1"
